### ISSUE TITLE
Subdirectories/folder feature

### ISSUE THREAD

**Issue by marcrleonard** [redimp/otterwiki#25]
Hello! I was curious if you had a plan to implement sub-directories/folders. I'd be happy to work on this. This morning I actually dove in to come up with a structure that could work:
<img width="943" alt="image" src="https://user-images.githubusercontent.com/24658895/194312813-43d0fcf0-9353-4520-82d0-faa7251fb559.png">
This will yield:
```
respoitory/
    home.md
    a_folder/
        a_folder.md
```
Navigating to the URL of `/a_folder` will yield `/a_folder/a_folder.md` (kind of like it's index.html page). This works from some simple matching logic to check if the pagepath  you're trying to access is a folder or a page. This moves the 'storage' checking aspect upfront (before crafting the paths on disk).

<img width="783" alt="image" src="https://user-images.githubusercontent.com/24658895/194314193-6e213286-4900-41e5-9f0c-90f18c2a402c.png">

Let me know if you'd like me to proceed with a PR.

**Comment by marcrleonard** [comment]
I went ahead and issued a PR. I apologize upfront as I accidentally included the other features in this branch so that may make it tricky to view. Luckily, not too many files changes.

https://github.com/redimp/otterwiki/pull/26

**Comment by marcrleonard** [comment]
I'm still working on this. I put a brief pause to work on a TOC mechanism to help test that the sub folders would be working properly. The TOC would help with the tests, and be a nice view for the user (just submitted it in the issues)

**Comment by marcrleonard** [comment]
Ok, this appears to be working well. The user experience is so-so (see points below), but the inner working on the sever side work. A few things to note:

1. You established using the "/Home" path as an 'index' page of the root of the app. I've implemented the same paradigm for folders. So if you attempt to access a folder "/My_Folder/" under the hood it will look for `/My_Folder/Home.md`
2. At **_first_**, I thought we would want to block creating a folder that has the same name as a file (which is it's attachments folder name) Example: If a user has `/Wiki Page.md` they could not create `/Wiki Page/Home.md` ... But I realized this would be somewhat of a false constraint. It does _slightly_ muddy the filesystem as attachments and pages would share folders, but when I really thought about it, it didn't really bother me. Anyways, that is something to consider. (sidenote: makes me wonder if attachments even need to be in a sub folder?)
3. With these changes, the app can use full paths to craft files/folders. For instance, on the create page, you can enter "some folder/another folder/a page" and that will create the nested folders (`some folder/`, `another folder/`) and the page inside. I'm not sure this is 100% clear to the user, but it definitely makes the creation of folders/files powerful. 
4. On the create page, I added a 'create as folder' switch (like the screenshot above), but actually, I think this is confusing. I'd rather just put a note in the ui that if you'd like to create a folder, just end the page name with a `/` ie. "some folder/" ... and under the hood this will create `/some folder/Home.md` ... Thoughts?

I have not yet issued a PR as I wanted to get your thoughts on these points. Let me know what you think!

**Comment by redimp** [comment]
First some thouhts to your points (in an order that helps following my thoughts):

- 1.) `Home` is only meant to be the "index" of the root of the repository, for the subdirectories, see below.

- 3.) I like the idea of using full paths and folders. If you need structure I would say simple but powerful.

- 4.) I think letting the user enter the full path (maybe with examples?) is a usage pattern that is well known and easy to understand.

- 2.1)  I have a simple propsal for the problem: What is the index of a subfolder? The parent file.

With a repository like
```
home.md
subspace.md
subspace/content.md
```

- a) `https://example.wiki/` leads to `home.md`
- b) `https://example.wiki/subspace` leads to `subspace.md` 
- c) `https://example.wiki/subspace/` leads to `subspace.md`  
- d) `https://example.wiki/subspace/content` leads to `subspace/content.md`

I dont see any problem of the `.md` files sharing the folder with the attachments. If a `.md` is supposed to be an attachment it can be viewed in the wiki ... but we would have to add a download functon for `.md` files though.

a), b) and d) are what is state of the current features. I forgot that d) was already implemented. I just tested it:

```
$ touch subspace.md
$ mkdir subspace
$ touch subspace/example.md
$ git add subspace*
$ git commit -m"added subspace"
```

open http://localhost:8080/subspace/content

<img width="452" alt="image" src="https://user-images.githubusercontent.com/112056/196785490-a02690c9-d5c8-4015-9644-7ef0678c0a23.png">

- 2.2) If someone deletes `/subspace` that would delete the `subspace.md`, all the attachments and of course all the pages in `subspace`. Thats in my opinion logical. (I already thought about this when I added the warning in 0774329

**Comment by marcrleonard** [comment]
Ok, lets try that and see how it feels ergonomically. I do think it's a little quirky that while accessing a folder-esque path (like `https://example.wiki/subspace/`) would produce `subspace.md` purely given how filesystems and web pages work (under the hood of a web server that would probably look for `/subspace/index.html`), but lets give it a shot. That will be easy to do on my current branch. 

I will incorporate the changes you've outlined and issue a PR.

**Comment by redimp** [comment]
The webserver assumes nothing when `https://example.wiki/subspace/` is fetched. Everything but `/static` is proxied without question to flask / the otterwiki app. So it's purly our problem ;)

Looking forward to your PR!

**Comment by marcrleonard** [comment]
In our case, yes. But I mean in conventional web server/filesystem access, it works kind of like I outlined. Try this example:

```
mkdir testserver
cd testserver 
echo "toplevel" > index.html
mkdir sublevel
echo "sublevel" > sublevel/index.html
python -m SimpleHTTPServer 8000
```

When you go to `localhost:8000/` it will yield "toplevel" text
When you go to `localhost:8000/sublevel`it will yield "sublevel" text

So essentially follow the pattern of a particular 'folder' route looking for a default page (index.html) in that directory.

**Comment by marcrleonard** [comment]
PR here: https://github.com/redimp/otterwiki/pull/42

**Comment by redimp** [comment]
This looks very promising! I will merge later today, we are traveling, so I might have time to check for bugs. Already found a bug in the search, which is not linking into the correct subdirectory/page when found.

**Comment by redimp** [comment]
Merged the PR in 481f2add6fa16abda60ea7c9a2593116367dd1a8.

Added pages in subdirectories to the PageIndex and fixed the Search so that it works as intended.
